### DISCUSSION TITLE
Feature Request: Ability to Map a String

### DISCUSSION THREAD

**Discussion by Ktedon** [vlang/v#14626] [Ideas]
I just started building a compiler for a custom dialect of lambda-prolog in Vlang. When trying to build an identifier parser, I tried mapping each character to whether it could at the beginning of an identifier, if it could be in the tail, and if it would terminate the identifier like this:
```v
identifier_map := program[0..19].map(
    match it {
	`A`...`Z` { i8(0) }
	`a`...`z` { i8(0) }
	`œ`...`Ÿ` { i8(0) }
	`À`...`Ö` { i8(0) }
	`Ø`...`ö` { i8(0) }
	`ø`...`ÿ` { i8(0) }
	`_`       { i8(0) }
	`š`       { i8(0) }
	`ª`       { i8(0) }
	`µ`       { i8(0) }

	`×`       { i8(1) }
	`÷`       { i8(1) }
	`0`...`9` { i8(1) }
	`~`...`™` { i8(1) }
	`¡`...`©` { i8(1) }
	`«`...`´` { i8(1) }
	`¶`...`¿` { i8(1) }
	else { i8(2) }
    }
)
```
Note: the program variable is a string.

When I try to compile this I get: `error: unknown method or field: string.map`.

If strings are just read-only array, why can't I use array methods on strings? It makes sense IMO.

**Comment by JalonSolov** [comment]
A V string is not an array.  It can sort of act like one, but the declaration is
```v
pub struct string {
pub:
	str &byte = 0 // points to a C style 0 terminated string of bytes.
	len int   // the length of the .str field, excluding the ending 0 byte. It is always equal to strlen(.str).
	// NB string.is_lit is an enumeration of the following:
	// .is_lit == 0 => a fresh string, should be freed by autofree
	// .is_lit == 1 => a literal string from .rodata, should NOT be freed
	// .is_lit == -98761234 => already freed string, protects against double frees.
	// ---------> ^^^^^^^^^ calling free on these is a bug.
	// Any other value means that the string has been corrupted.
mut:
	is_lit int
}
```
So it's just a "string" of bytes plus a length and a flag for the type.

**Comment by Ktedon** [comment]
It says in the documentation: `In V, a string is a read-only array of bytes.` Maybe that's something that should be fixed.

**Comment by JalonSolov** [comment]
Any sequential memory block could be considered "an array of bytes".  So both the documentation and the actual implementation are technically correct.

However, it is _not_ a V array, which has a different structure.

That said, someone could probably create a `.map` method on string if they really wanted to do so.  It isn't necessary, since you can iterate the bytes in the string instead of introducing a new method.  The new method would just be syntactic sugar.

**Comment by metif12** [comment]
first get the bytes or runes array of string then map and convert result back to string.
```v
identifier_map := program[0..19].bytes().map(map_fn).bytestr()
```
